pest/src/inputs_private/input.rs

```diff
@@ -0,0 +1,83 @@
+// pest. The Elegant Parser
+// Copyright (C) 2017  Dragoș Tiselice
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::fmt::Debug;
+use std::ops::Range;
+
+/// A `trait` that defines an input for a `Parser`. It should be implemented by custom input sources
+/// as minimally and as efficiently possible. `unsafe` methods should not be called directly; in
```

Comment:
I've added a comment clarifying the motivation behind. I'm not completely convinced that this is the best method to handle things, however.